ortb/clone_test.go

```diff
@@ -536,6 +536,326 @@ func TestCloneUser(t *testing.T) {
+		assert.NotSame(t, given, result, "pointer")
+	})
+}
+
+func TestCloneBrandVersionSlice(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneBrandVersionSlice(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{}
+		result := CloneBrandVersionSlice(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("one", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{
+			{Brand: "1", Version: []string{"s1", "s2"}, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneBrandVersionSlice(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item-pointer")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item-pointer-ext")
+	})
+
+	t.Run("many", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{
+			{Brand: "1", Version: []string{"s1", "s2"}, Ext: json.RawMessage(`{"anyField":1}`)},
+			{Brand: "2", Version: []string{"s3", "s4"}, Ext: json.RawMessage(`{"anyField":1}`)},
+			{Brand: "3", Version: []string{"s5", "s6"}, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneBrandVersionSlice(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item0-pointer")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item0-pointer-ext")
+		assert.NotSame(t, given[1], result[1], "item1-pointer")
+		assert.NotSame(t, given[1].Ext, result[1].Ext, "item1-pointer-ext")
+	})
+}
+
+func TestCloneBrandVersion(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneBrandVersion(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := &openrtb2.BrandVersion{}
+		result := CloneBrandVersion(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("populated", func(t *testing.T) {
+		given := &openrtb2.BrandVersion{
+			Brand:   "Apple",
+			Version: []string{"s1"},
+			Ext:     json.RawMessage(`{"anyField":1}`),
+		}
+		result := CloneBrandVersion(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+		assert.NotSame(t, given.Ext, result.Ext, "ext")
+	})
+
+	t.Run("assumptions", func(t *testing.T) {
+		assert.ElementsMatch(t, discoverPointerFields(reflect.TypeOf(openrtb2.BrandVersion{})),
+			[]string{
+				"Version",
+				"Ext",
+			})
+	})
+}
+
+func TestCloneSource(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneSource(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := &openrtb2.Source{}
+		result := CloneSource(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("populated", func(t *testing.T) {
+
+		given := &openrtb2.Source{
+			FD:     1,
+			TID:    "Tid",
+			PChain: "PChain",
+			SChain: &openrtb2.SupplyChain{
+				Complete: 1, Nodes: []openrtb2.SupplyChainNode{{ASI: "asi", Ext: json.RawMessage(`{"anyField":1}`)}}},
```

Comment:
Did you want the populated ext to be `openrtb2.SupplyChain.Ext` instead of `openrtb2.SupplyChainNode.Ext`?

Reply:
Yes, refactored